### PR TITLE
Write block entity nbt on networking thread instead of server one

### PR DESCRIPTION
**Pull request by Patbox**
This pr adds ServerPacketWriter, a utility interface for creation on CustomPayloadS2CPackets that write on networking thread. The implementation is lifted directly with polymer with few simplifications, since you don't need player context or packet version handling.

From my testing (within polymer) it indeed is faster (with most noticeable effect with large, nbt driven packets).

Ran this in singleplayer and all methods wrote and read data correctly

### CODE REVIEW

File: src/main/java/net/modfest/fireblanket/mixin/MixinCustomPayloadS2CPacket.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    @Inject(method = "write", at = @At("TAIL"))
+    private void fireblanket$write(PacketByteBuf buf, CallbackInfo ci) {
+        if (this.writer != null) {
+            this.writer.write(buf, this.channel);
+        }
+    }
+
+    /*
+    * This section is a bit of a "hack" to make sure it behaves more or less the same on singleplayer as on server.
+    * Battle tested in polymer™
```
Comment (jaskarth) [inline src/main/java/net/modfest/fireblanket/mixin/MixinCustomPayloadS2CPacket.java]:
```suggestion
```
Explain the "hack" here more

File: src/main/java/net/modfest/fireblanket/mixin/MixinCustomPayloadS2CPacket.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            return Unpooled.buffer();
+        }
+        return vanilla;
+    }
+
+    @Environment(EnvType.CLIENT)
+    @Inject(method = "getData", at = @At("RETURN"))
+    private void fireblanket$writeData(CallbackInfoReturnable<PacketByteBuf> cir) {
+        if (this.writer != null) {
+            var buf = cir.getReturnValue();
```
Comment (jaskarth) [inline src/main/java/net/modfest/fireblanket/mixin/MixinCustomPayloadS2CPacket.java]: Avoid var (goes for everywhere)

File: src/main/java/net/modfest/fireblanket/net/writer/ServerPacketWriter.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import net.minecraft.util.Identifier;
+
+
+/**
+ * ServerPacketWriter is a utility method for creating CustomPayloadS2CPackets that write on network thread
+ * instead of server one.
+ *
+ * This is a (minimally) simplified version of interface from Polymer Networking API.
+ * In case of bigger data (like polymer's registry sync), performance increase was noticeable,
+ * so I thought it would be good idea to bring it here.
```
Comment (jaskarth) [inline src/main/java/net/modfest/fireblanket/net/writer/ServerPacketWriter.java]: Avoid first person in comments

File: src/main/java/net/modfest/fireblanket/net/BundledBlockEntityUpdatePacket.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
+package net.modfest.fireblanket.net;
+
+import net.minecraft.network.PacketByteBuf;
+import net.minecraft.registry.Registries;
+import net.minecraft.util.Identifier;
+import net.modfest.fireblanket.net.writer.ServerPacketWriter;
+
+
+public record BundledBlockEntityUpdatePacket(BEUpdate[] updates) implements ServerPacketWriter {
```
Comment (jaskarth) [inline src/main/java/net/modfest/fireblanket/net/BundledBlockEntityUpdatePacket.java:9]: Does this need to be an array? If you keep the list you can remove the `toArray()` above, reducing an allocation

**Review by jaskarth — CHANGES_REQUESTED** [review]
Thanks! I have a few comments.

File: src/main/java/net/modfest/fireblanket/net/BundledBlockEntityUpdatePacket.java (hunk 4 above)
Comment (Patbox) [reply src/main/java/net/modfest/fireblanket/net/BundledBlockEntityUpdatePacket.java:9]: List was later cleared. Through using old list here and creating new one is also an option.

File: src/main/java/net/modfest/fireblanket/net/writer/ServerPacketWriter.java (hunk 3 above)
Comment (Patbox) [reply src/main/java/net/modfest/fireblanket/net/writer/ServerPacketWriter.java]: I guess my comment style doesn't fit here. Will change that :p

File: src/main/java/net/modfest/fireblanket/net/BundledBlockEntityUpdatePacket.java (hunk 4 above)
Comment (jaskarth) [reply src/main/java/net/modfest/fireblanket/net/BundledBlockEntityUpdatePacket.java:9]: Ah, right. The array solution is probably fine here then.

**Comment by Patbox** [comment]
Applied all suggestions

**Review by jaskarth — APPROVED** [review]
Looks good, thanks for this!

**Pull request merged**